make gfx.show customizable

FirefoxMetzger opened the pull request:
Closes: https://github.com/pygfx/pygfx/issues/378 and https://github.com/pygfx/pygfx/issues/383

This PR adds more flexibility to `gfx.show` by allowing the replacement of various parts of this default setup. In particular, it allows using a custom: `canvas`, `event_loop`, `renderer`, `controller`, `camera`, and `draw_function`. It also enhances the default draw function by introducing two callbacks: `before_render` and `after_render`, which allow customizing the draw call (eg. to add animations) without having to replace the entire thing.

I also changed the way lighting is treated by `gfx.show` and only add a light source when the user does not provide a `Scene` object. My angle here is that, if the user provides an entire scene, they probably have their reasons and don't want us to mess with it. I'm happy to revert this though if you think it would be better to add light.

I'm also not super happy with always having a controller (what if a person wants none), but I think this is a minor issue.

Finally, a point that we could consider is if we want to change the callbacks from kwargs on `show` to decorators. Ie., if we would like to enable something like:

```python
import pygfx as gfx

@gfx.show.before_render
def animate():
    ...

gfx.show()  # uses animate
```

---

Since this is my first PR that contributes actual code and I didn't find any active coverage tracking, I will ask explicitly: @almarklein  @Korijn What are your expectations towards tests and coverage?

Korijn commented on pygfx/utils/show.py:
```diff
@@ -34,33 +44,47 @@ def show(object: WorldObject, up=None):
 
-    scene.traverse(check_light, False)
-    if not light_count:
-        camera.add(DirectionalLight(0.8))
```
I think this specific line has gone missing?

Korijn commented:
The improvements here are very nice and I think for now this will do.

I propose to keep track of your idea for a decorator and dataclass style implementation in a new ticket so we can follow up in the future.

Korijn commented:
Hmm, looks like test_examples.py doesn't support gfx.show yet. I was a little concerned that would be the case... Should be doable though. I'll check it out tomorrow.

FirefoxMetzger commented:
I will also have to write a decent docstring for the new function and then there are unit tests ... 😅

almarklein commented on pygfx/utils/show.py:
```diff
@@ -34,33 +44,47 @@ def show(object: WorldObject, up=None):
+        canvas = WgpuCanvas()
+        event_loop = run
+    elif event_loop is None:
+        # TODO: find the matching run function
```
This should do the trick:
```py
sys.modules[renderer.target.__module__].run
```

almarklein replied:
Do you think this would mean we can drop the `event_loop` argument (since it comes with the canvas)? If not I propose a different name for `event_loop`, because it sounds like an object, which may be confusing.

FirefoxMetzger replied:
If we can find the `run` function for a given canvas then that would make it easier indeed. The question is if we want to allow people to experiment with their own event loop for a canvas, or if this is out of scope for now.

If it is out of scope, we can drop `event_loop` for now.

Korijn replied:
Agree with dropping `event_loop` for now

almarklein commented on pygfx/utils/show.py:
```diff
@@ -34,33 +44,47 @@ def show(object: WorldObject, up=None):
+
+        def animate():
+            if before_render is not None:
+                before_render()
```
If we'd provided `renderer` as an argument to `before_render` and `after_render` we could e.g. request new draws in the user code.

FirefoxMetzger replied:
Would that be an advantage? The current default loop already requests a new draw at the end in any case, so I am inclined to say this would be the realm of using a custom `event_loop`, which could totally have a reference to the canvas and the renderer.

Korijn replied:
I would recommend to see if the need comes up while you are rewriting more examples later on.

almarklein commented on pygfx/utils/show.py:
```diff
@@ -34,33 +44,47 @@ def show(object: WorldObject, up=None):
 
         background = Background(None, BackgroundMaterial((0, 1, 0, 1), (0, 1, 1, 1)))
         scene.add(background)
+        scene.add(AmbientLight(0.2))
```
I prefer the old version of determining whether to add lights. Imagine:
```py
cube = gfx.Mesh(...)
gfx.show(cube)
```
Works fine, now I modify my example to show two cubes:
```py
scene  = gfx.Scene()
scene.add(gfx.mesh(...))
scene.add(gfx.mesh(...))
gfx.show(scene)
```
and now it's all dark :)

Korijn replied:
You could use a Group, but I agree that it's maybe surprising behavior

FirefoxMetzger replied:
The opposite holds true as well though "The scene works in `gfx.show`. Why don't I see anything in my actual setup!?".

It's meaningful to add lights if we only get an object and construct a default scene (users will likely place that object into their own scene later). However, when we get a full scene I think we should be mindful before modifying it; especially when we think of `gfx.show` as a debugging tool.

Korijn replied:
How about just adding a kwarg `add_lights=True`

almarklein replied:
> The opposite holds true as well though "The scene works in gfx.show. Why don't I see anything in my actual setup!?".

Good point. There is a chance of confusion either way.

What bothers me about `if isinstance(object, Scene)` is that it means the behavior of this function depends on whether a single object is passed, compared to multiple objects grouped in a scene. E.g. adding lights and background.

After some thought, I like @Korijn's suggestion about using `Group`: we can keep the behavior, but add a note in the docstring about the behavior when a `Scene` is given, and that a `gfx.Group()` should probably be used for multiple objects. And also use that in examples.

To take this idea further, if we create the camera, it should only be added to the scene (with a light attached) when we also create the scene here.

almarklein replied:
I added a suggestion for my last point. Could you update the docs to explain `Scene` vs `Group`?

Korijn replied:
I really do think an `add_lights=True` kwarg would be better, honestly. It's explicit, and obvious.

almarklein replied:
One reason I was leaning towards using the type of the given object to decide whether to add stuff, was that we also background and a camera to it. Would it make sense to generalize  `add_lights=True` to something like `modify_scene=True`?

Korijn replied:
How about: `add=('lights', 'camera', 'background', 'controller')`

I wish I could put `action` in there :)

almarklein replied:
mmm, what about `add_lights` and `add_background` args? (The camera being added is by virtue of adding the directional light.)

Korijn replied:
They're all equally acceptable to me, I just think it's better to lean towards controlling the behavior of `gfx.show()` via explicit keyword arguments than letting it be derived from the scene graph in a rather opinionated manner.

FirefoxMetzger replied:
Hm. At first, I was going to agree with `add_lights`, but when thinking about it a bit more I am not sure anymore if it solves the problem. 

Assuming `add_lights` has a default value like `add_lights=True` then chances are most people will use `gfx.show(scene)` and not know/care about most additional kwargs unless explicitly prompted to check. Especially if there are many. (In my experience, many devs only check the docs if something is broken and then only if SO doesn't have good suggestions.) So even though we can now turn this behavior off, it is still somewhat unintuitive and we are not exactly advertising it.
As such, the switch itself doesn't really solve the "Why does it work in `gfx.show`, but not in my application" problem.

We could, ofc, choose `add_lights=False` as a default but that would defeat the point of the kwarg. It's equal work to do `scene.add(gfx.SomeLight())` vs explicitly passing `add_lights=True`, so there is nothing gained. 

Also, in the case of us wrapping the input in a scene, we will likely have to do the lights and ignore this kwarg. This would make our kwargs quite complicated, because we already have `canvas`, `before_render`, and `after_render` that may be ignored in certain circumstances. That's quite a complex API for something that is meant to be a "quick scaffolding tool to debug your application or develop pygfx".

FirefoxMetzger replied:
Thinking of `gfx.show` as a diagnostics/debugging tool, I like the following:

If we don't get a `Scene`, we create one and add a light since this scene typically contains the ambient light. If we do get a `Scene`, we check if there are light sources and if not, we show a warning along the lines of `"Your scene does not contain a light source and will appear black."`. 

This way, there are no surprises. We can visualize single objects as expected, we don't modify existing scenes, and we give useful feedback on "ad-hoc" scenes like the one from @almarklein 's description. Would this work for you @almarklein ?

almarklein replied:
Yeah, I think this could work 👍 . I'd say something like "... and some objects may not be visible".

FirefoxMetzger replied:
This is now implemented :)

almarklein commented on pygfx/utils/show.py:
```diff
@@ -34,33 +44,47 @@ def show(object: WorldObject, up=None):
-    if not light_count:
-        camera.add(DirectionalLight(0.8))
-        scene.add(AmbientLight(0.2))
+    if renderer is None:
+        renderer = WgpuRenderer(canvas)
```
A custom canvas should be disallowed if a custom renderer is given. We don't really need the canvas here, we can do the `request_draw()` call below on the renderer instead.

FirefoxMetzger replied:
Sure.

almarklein commented on pygfx/utils/show.py:
```diff
@@ -34,33 +44,47 @@ def show(object: WorldObject, up=None):
 
-    canvas.request_draw(animate)
-    run()
+    canvas.request_draw(draw_function)
```
```suggestion
    renderer.request_draw(draw_function)
```

FirefoxMetzger replied:
I would just get the `canvas = renderer.target`. This makes the code overall cleaner :)

almarklein reviewed (changes requested):
This is becoming an all-round powertool!

almarklein commented on pygfx/utils/show.py:
```diff
@@ -11,56 +14,104 @@
-    look_at = camera.show_object(object)
-    scene.add(camera)
+    if renderer is not None:
+        canvas = renderer.target
```
```suggestion
        if not (canvas is None or canvas is renderer.target):
            raise ValueError("The given canvas does not match the given renderer.")
        canvas = renderer.target
```

almarklein commented on pygfx/utils/show.py:
```diff
@@ -11,56 +14,104 @@
-    canvas.request_draw(animate)
-    run()
+    canvas.request_draw(draw_function)
+    sys.modules[renderer.target.__module__].run()
```
Use Canvas like in the line above.
```suggestion
    sys.modules[canvas.__module__].run()
```

almarklein commented on pygfx/utils/show.py:
```diff
@@ -11,56 +14,104 @@
+
+    if not camera:
+        camera = PerspectiveCamera(70, 16 / 9)
+        camera.add(DirectionalLight(0.8))
+        scene.add(camera)
```
```suggestion
        if scene is not object:
            camera.add(DirectionalLight(0.8))
            scene.add(camera)
```

FirefoxMetzger replied:
How would this work for situations where we pass a scene to the renderer but don't pass a camera?

almarklein replied:
Then we create a camera, but don't add it to the scene. The camera does not have to be in the scene for it to work. The reason it is added to the scene in these lines, is to make the light source move along with it (as if you had a head-light).

FirefoxMetzger replied:
I wasn't quite sure how to address this comment, especially after the reactor. Since I'm not sure what we were after here, I can't really say if this is resolved now or still something to improve. @almarklein could you have a look and let me know if we can resolve this?

Korijn commented:
So it seems to support gallery generation, we need to expose some kind of reference from each example module to the renderer and/or canvas.

That implies we need something to hold on to... like a `gfx.Display` class. :) I did not have this insight earlier, it's costing you time and rework, I'm sorry for that.

FirefoxMetzger commented:
> So it seems to support gallery generation, we need to expose some kind of reference from each example module to the renderer and/or canvas.

The exact requirement is that we need to decide if (and what) to draw based on the input that sphinx-gallery gives the scraper's callback. Those are:

- `block`, a 3-tuple with `(block_type, block_source, starting_lineno)` where `block_type="code"` and `block_source` is the raw source code of that block (for us the entire example)
- `block_vars` a dict where the relevant fields are `src_file` (path to file), `example_globals` (global variables), `fake_main` (variables defined in the main namespace). The variables are collected by running the example up to and including the current block.

I can make the gallery work as long as there is a way for me to inspect the above (or the global state of pygfx and wgpu) and check (1) if I should create an image (or GIF) and (2) how to generate the image.

Currently, this is done by checking for a global `gfx.Camera` and `gfx.Scene` and if both are defined I create a canvas and a renderer and take a snapshot. This is very hacky though, so I am hoping to find a better solution, especially in light of `gfx.show` hiding `gfx.Camera` in most examples.

> That implies we need something to hold on to... like a gfx.Display class. :) I did not have this insight earlier, it's costing you time and rework, I'm sorry for that.

@Korijn I didn't either. We started this discussion before I started writing a custom scraper for sphinx-gallery, so we could not have known of this requirement at that stage. What's I think more interesting/important is how we best proceed from here.

I think there are three ways forward:

- I change this PR, introduce `Display`, and use it to build the gallery
- We introduce something like [`plt.get_fignums()`](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.get_fignums.html) in pygfx to track canvases. (The docs are not very extensive. Essentially, it tracks currently open figures so that you can find and access them)
- We find a feature in `wgpu` that can check for currently existing canvases

>

Korijn commented:
I'm not a fan of tracking things globally.

> I change this PR, introduce Display, and use it to build the gallery

This seems most sensible.

almarklein commented:
Once we have `Display` as an alternative to `gfx.show()`, we can categorize all examples in three groups:
* Examples that use `Display`. In the gallery-builder we can look for the object and use it.
* Examples that do not use display, but do have a `renderer`. In the gallery-builder we can use that.
* Examples that do more custom stuff - we do not include these in the gallery.

FirefoxMetzger commented:
Alright. Let me get this branch up to data and then we see how it looks 🚀

FirefoxMetzger commented:
That turned out to take longer than I expected. I got a little sidetracked trying to find a good way to implement the getter idea from https://github.com/pygfx/pygfx/issues/383. With the current changes, the cube example simplifies to:

```python
import pygfx as gfx

cube = gfx.Mesh(
    gfx.box_geometry(200, 200, 200),
    gfx.MeshPhongMaterial(color="#336699"),
)


@gfx.Display.before_render
def animate():
    rot = gfx.linalg.Quaternion().set_from_euler(gfx.linalg.Euler(0.005, 0.01))
    cube.rotation.multiply(rot)


if __name__ == "__main__":
    gfx.show(cube)

```

(Although we probably want to use `disp = gfx.Display(); disp.show()` in the examples to allow the gallery to pick up the Display object and render it.)

I've done the following changes to the PR:

- I introduced `Display()` which now holds all the data for the setup and made `gfx.show` a thin wrapper around it.
- I introduced a Descriptor class that allows setting the default callback for Display using decorators (more below)
- I've overhauled all the docstrings in the class
- I went through all the open discussions and implemented the suggestions.

What I did not do is write additional tests, since the code will be tested by every example once I refactored them.

---

A note on the decorator-style callback: If you look closely, you will see that the decorator is called on the class-level (`@gfx.Display.before_render` vs `disp = gfx.Display() ... @gfx.disp.before_render`). This is because I couldn't figure out how to make `@gfx.disp.before_render` work directly. This decorator would call `gfx.disp.before_render.__get__` which will yield the currently set callback instead of the decorator function and I don't think we can work around this.

In leu of the preferred option, we can have the following three alternatives:

```python

#  Decorator on the class (<- I implemented this)
@gfx.Display.before_render
def animate():
    ...

#  Decorator on a different method
@disp.update.before_render
def animate():
    ...

#  Return a wrapper with a custom method
@disp.before_render.update
def animate():
    ...
```

The first is complementary to the other two (we can implement both in parallel), but I didn't do that, because the latter two look a little cryptic, which may not be super maintenance friendly. That said, if you like the API of those options I can implement these.

The downside of the class-level approach is that we don't have access to instance-level variables, so the best we can do is modify the default value of the callback. This is what the current implementation does and as such, there is a caveat in cases where we prepare multiple displays at the same time as they will all share this default value and multiple calls to `@gfx.Display.before_render` will overwrite each other. This should be rare though, since in most cases we will only have one display at the time, i.e., create a scene, display, close, next scene.

Korijn commented on pygfx/utils/show.py:
```diff
@@ -16,9 +17,243 @@
+            self.after_render = after_render
+
+    def default_draw(self):
+        if self.before_render is not None:
```
How can this ever be `None`? Doesn't it resolve to the class level RenderCallback always?

Korijn commented on pygfx/utils/show.py:
```diff
@@ -16,9 +17,243 @@
+        if value is not None:
+            return value
+
+        # persist callback to prevent it from changing
```
Can you explain how this prevents it? I'm not super clear on the control flow here

FirefoxMetzger replied:
Right. Under the hood, the descriptor uses a `_<name>` attribute on the instance to manage the actual callback, i.e., if we do

```python
disp = Display()

disp.before_render = animate
```

then the descriptor creates `disp._before_render` to store the callback on `disp`. If that value doesn't exist or is `None` we will fall back to `gfx.Display.before_render.default_callback`. This is what the `if value is None: ...` does.

This works for most cases, except when we have an already running `disp` (e.g. inside an earlier notebook cell) that uses the default callback. If, in this case, we use the `@gfx.Display.before_render` decorator while setting up the next display object we may modify the running one. Since the decorator sets the class-level default (which the previous `disp` is using) we risk the rendering loop of the running renderer changing. To prevent this interaction I take the value `Display.before_render.default_callback` during the first run and persist it on the instance by setting `disp._before_render = Display.before_render.default_callback`. Now future changes of the class-level default can no longer impact the instance-level `disp` once the display starts running/showing.

Writing this out makes two enhancements come to mind: (1) We probably want to not use `default_callback` if the user has explicitly set `disp.before_render = None`. (2) We could add a `Display.finalize()` function that does all the setup that can be done without making a canvas pop up. I'm thinking in particular of setting all the callbacks and making them independent of the class-level defaults. The second one is not needed right now, but is quite easy to incorporate.

Korijn replied:
👍 thanks for explaining

FirefoxMetzger replied:
No problem. Should I add a clarifying comment in the source for future reference somewhere? (I also did a refactor of this section just now, so it might be a little more intuitive now)

Korijn commented on pygfx/utils/show.py:
```diff
@@ -16,9 +17,243 @@
+            scene.add(AmbientLight(), DirectionalLight())
+        self.scene = scene
+
+        if len(self.find_children(Light)) == 0:
```
Could you use `any()` here? That would enable the more efficient generator implementation you hinted at

FirefoxMetzger replied:
hm, as long as `bool(gfx.Light(...)) == True` this should be possible. Its a bit weird to think of the boolean-ness of a WorldObject, but I like how `if any(self.find_children(Light))` looks syntax-wise 🤩 I will change it.

Korijn commented on pygfx/utils/show.py:
```diff
@@ -16,9 +17,243 @@
+            self.camera = PerspectiveCamera(70, 16 / 9)
+            self.camera.add(DirectionalLight())
+            # do we really want to modify the scene and add
+            # a camera?
```
If you don't do this, the directional light is not in the scene either :)

FirefoxMetzger replied:
Yes. my question is more from the angle of "do we want to modify the input scene?". It's similar to our discussion about adding lights by default: How worried are we about the side effects of permanently modifying the user's scene?

Korijn replied:
Yes, well, I think we had that discussion earlier, I think all proposals are still on the table. Let's just go with this approach for now and adjust it later if we think of something better.

Korijn commented on pygfx/utils/show.py:
```diff
@@ -16,9 +17,243 @@
+        else:
+            pass
+
+        controllers = self.find_children(Controller)
```
I don't think Controllers are a WorldObject subclass, so you will never find them in the scene

FirefoxMetzger replied:
Oh, that's true. In that case I will drop this logic. @Korijn Do you know if there is another way to discover existing controllers, or will it be sufficient to only use one if passed to the Display?

Korijn replied:
I don't think there is another way, no!

FirefoxMetzger replied:
Then we go with explicitly passing one until a better option develops :)

Korijn commented on pygfx/utils/show.py:
```diff
@@ -16,9 +17,243 @@
+            self.controller.add_default_event_handlers(self.renderer, self.camera)
+
+        self.canvas.request_draw(self.draw_function)
+        sys.modules[self.canvas.__module__].run()
```
I think we should open a ticket on wgpu-py to provide a simple reference, something like `WgpuCanvas.run`

FirefoxMetzger replied:
See: https://github.com/pygfx/wgpu-py/issues/307

Korijn commented on pygfx/utils/show.py:
```diff
@@ -16,9 +17,243 @@
+        self.canvas.request_draw(self.draw_function)
+        sys.modules[self.canvas.__module__].run()
+
+    # this should probably live on WorldObject
```
Agreed

FirefoxMetzger replied:
@Korijn Should this happen in a separate PR where we move the function over, or do you want this to be part of this PR?

Korijn replied:
Let's create a ticket and follow up later

FirefoxMetzger replied:
See https://github.com/pygfx/pygfx/issues/385

Korijn commented on pygfx/utils/show.py:
```diff
@@ -16,9 +17,243 @@
+        objects = list()
+
+        # can we make traverse a generator?
+        # [x for x in self.scene.traverse(filter=lambda x: isinstance(x, clazz))
```
`traverse` is more like `map`, it's purpose is to apply a function to all objects in a (sub)graph. I like your suggestion but it has a different purpose. You can just add a new function to `WorldObject`, I like the name `find_children`, could also be `query`, I don't mind.

FirefoxMetzger replied:
Okay. In that case, we can just add a `find_children` function. 

The prototype I had in mind was something like this:

```python
def traverse(self, filter:Callable[[WorldObject],bool], skip_invisible=False):
    if skip_invisible and not self.visible:
        return

    if not filter(self):
        return

    yield self

    for child in self._children:
        yield from child.traverse(callback, skip_invisible)
```

With the above you can also do modifications like:

```python
# set all light intensities to .8
for light in scene.traverse(filter=lambda x: isinstance(x, gfx.Light)):
    light.intensity = .8
```

So maybe the name `iter` or `iter_children` would make more sense?

Korijn replied:
Nice, I like it! I would say just `iter` is a very nice fit.

Korijn commented:
Really nice work! 👍 

`@disp.before_render` if this can't be done, that too bad. I wouldn't worry about having a decorator feature

FirefoxMetzger commented:
> @disp.before_render if this can't be done, that's too bad. I wouldn't worry about having a decorator feature too much, it's not really critical to making this work.

How do you think of `@gfx.Display.before_render`? If you think it's too much/weird I can also kick it out. In that case we can remove the entire descriptor class (`RenderCallback`) and just use standard assignments.

Korijn commented:
I am personally leaning towards kicking it out if the examples still look nice without them :)

FirefoxMetzger commented:
@Korijn 

With `@gfx.Display.before_render`:

```python
import pygfx as gfx

cube = gfx.Mesh(
    gfx.box_geometry(200, 200, 200),
    gfx.MeshPhongMaterial(color="#336699"),
)


@gfx.Display.before_render
def animate():
    rot = gfx.linalg.Quaternion().set_from_euler(gfx.linalg.Euler(0.005, 0.01))
    cube.rotation.multiply(rot)


if __name__ == "__main__":
    disp = Display()
    disp.show(cube)
```

Without `@gfx.Display.before_render`:

```python
import pygfx as gfx

cube = gfx.Mesh(
    gfx.box_geometry(200, 200, 200),
    gfx.MeshPhongMaterial(color="#336699"),
)


def animate():
    rot = gfx.linalg.Quaternion().set_from_euler(gfx.linalg.Euler(0.005, 0.01))
    cube.rotation.multiply(rot)


if __name__ == "__main__":
    disp = Display()
    disp.before_render = animate
    disp.show(cube)
```

Korijn commented:
If it simplifies the implementation a lot, I would say lose the decorator support for now.

FirefoxMetzger commented:
And the decorator is gone :) Now it's only @almarklein 's comment and my question to @almarklein in the docstring of `Display`.

@Korijn Thanks for reviewing this on a weekend, btw 🚀 At the same time, please don't feel pressured by me working on the weekend. My PhD has ruined my working habits, but that doesn't mean that there are any expectations towards getting things reviewed over the weekend.

almarklein commented on pygfx/utils/show.py:
```diff
@@ -11,56 +15,229 @@
+            self.camera = existing_cameras[0]
+        elif not self.camera:
+            self.camera = PerspectiveCamera(70, 16 / 9)
+            self.camera.add(DirectionalLight())
+            self.scene.add(self.camera)
```
I saw that the two of you already had a discussion about this, but I think the docs are pretty clear that if a `Scene` is provided, that we don't add lights, so let's be consistent. @Korijn ?
```suggestion
            if object is not scene:
                self.camera.add(DirectionalLight())
                self.scene.add(self.camera)
```

FirefoxMetzger replied:
I will resolve https://github.com/pygfx/pygfx/pull/380#discussion_r1019023447 in favor of this discussion, since it is the same suggestion :)

Korijn replied:
Yes, this is a good suggestion.

I would recommend to evaluate the condition in a single place though. For example you could put `add_lights = not isinstance(object, Scene)` near the top and re-use that condition in both places where lights are added.

FirefoxMetzger replied:
Refactored.

almarklein commented on pygfx/utils/show.py:
```diff
@@ -11,56 +15,229 @@
+    after_render=None,
+    draw_function=None,
+):
+    """Display a WorldObject
```
This docstrings needs the "upstream" changes too :)

almarklein commented on pygfx/utils/show.py:
```diff
@@ -11,56 +15,229 @@
+        return objects
+
+
+def show(
```
I wonder if whether we'd ever use this, since we've found that we need that `display` object to be able to pick it up in the gallery and tests ...

almarklein replied:
Still nice to have I guess, and does not look like a maintenance burden, so let's just keep it.

almarklein reviewed (commented):
This looks nice!

Not a fan of how that decorator looked, plus decorators may be intimidating for Python novices, so I'm glad you removed it.

My comments are mostly nitpicks in docstrings. Except for the one where the camera is added (or not added) to the scene.

FirefoxMetzger commented:
@Korijn @almarklein This resolves all the currently open conversations. I also have nothing else to add, so if the tests are happy this one is RTR from my side :)

Korijn commented:
Nice work! You can merge as far as I am concerned. For some reason CI was not triggered though. It may require another git push. :/
